### PR TITLE
Changes for issue with ChecksumAddress

### PR DESCRIPTION
**Pull request by zacharyestep**
## Summary

Explain the goal of this change and what problem it is solving.

https://github.com/algorand/go-algorand/issues/51

## Test Plan

Run unit tests

### CODE REVIEW

File: cmd/algorelay/eb/eb.go [hunk 1]
```diff
@@ -18,9 +18,9 @@ package eb
 
 // Relay represents the configuration data necessary for a single Relay
 type Relay struct {
-	ID                    int64  // db key injected when loaded
```
Comment (algobolson) [inline cmd/algorelay/eb/eb.go]: this isn't wrong, but it's irrelevant to the core of the change around checksum addresses so I'd just like to see a more minimal change without reformatting of random other parts of the code. also in cmd/algorelay/relayCmd.go

File: cmd/goal/account.go [hunk 2]
```diff
@@ -625,7 +625,7 @@ var installParticipationKeyCmd = &cobra.Command{
 	Run: func(cmd *cobra.Command, args []string) {
 		if !partKeyDeleteInput {
 			fmt.Println(
-`The installpartkey command deletes the input participation file on
+				`The installpartkey command deletes the input participation file on
```
Comment (algobolson) [inline cmd/goal/account.go]: don't reindent this

File: config/config.go [hunk 3]
```diff
@@ -229,7 +229,7 @@ func initConsensusProtocols() {
 		RewardUnit:                 1e6,
 		RewardsRateRefreshInterval: 5e5,
 
-		ApprovedUpgrades:     map[protocol.ConsensusVersion]bool{},
```
Comment (algobolson) [inline config/config.go]: another unneeded reformat

File: data/basics/address.go [hunk 4]
```diff
@@ -29,8 +29,8 @@ type (
 	Address crypto.Digest
 )
 
-// ChecksumAddress is a representation of the short address with a checksum
-type ChecksumAddress struct {
+// checksumAddress is a representation of the short address with a checksum
```
Comment (algobolson) [inline data/basics/address.go]: is there anything still referencing `checksumAddress` or can we just delete it?

File: data/basics/address.go [hunk 5]
```diff
@@ -87,15 +87,15 @@ func UnmarshalChecksumAddress(address string) (Address, error) {
...
+// String returns a string representation of checksumAddress
+func (addr *checksumAddress) String() string {
 	var addrWithChecksum []byte
 	addrWithChecksum = append(addr.shortAddress[:], addr.checksum...)
 	return base32.StdEncoding.WithPadding(base32.NoPadding).EncodeToString(addrWithChecksum)
 }
 
 func (addr Address) String() string {
-	return fmt.Sprintf("%v", crypto.Digest(addr))
+	return addr.getChecksumAddress().String()
```
Comment (algobolson) [inline data/basics/address.go]: move the code in here and delete `checksumAddress` ?

File: data/basics/address_test.go [hunk 6]
```diff
@@ -25,22 +25,23 @@ import (
 	"github.com/algorand/go-algorand/protocol"
 )
 
+/*
```
Comment (algobolson) [inline data/basics/address_test.go]: delete instead of comment out

File: libgoal/participation.go [hunk 7]
```diff
@@ -214,12 +214,12 @@ func (c *Client) InstallParticipationKeys(inputfile string) (part account.Partic
 		return
 	}
 
-        // After successful install, remove the input copy of the
```
Comment (algobolson) [inline libgoal/participation.go]: revert irrelevant reformatting

File: cmd/algorelay/eb/eb.go (hunk 1 above)
Comment (zacharyestep) [reply cmd/algorelay/eb/eb.go]: These changes are form gofmt I think. just pushed a change to these two - didn't spot any others changed in this way. 

V/r,

File: data/basics/address.go (hunk 4 above)
Comment (zacharyestep) [reply data/basics/address.go]: It's used internally by address.go and that's it. I considered moving all of the functionality into Address struct, but the breakdown of responsibilities between the two types is both reasonable and eloquent.  With the private type checksumAddress being responsible for calculating, validating checksums etc.

File: data/basics/address.go (hunk 5 above)
Comment (zacharyestep) [reply data/basics/address.go]: can do, I like the current breakdown since you can unit test checksumAddress on its own inside of the basics package.

**Comment by zacharyestep** [comment]
> Thanks! This is a good start to cleaning up the Address/ChecksumAddress schism. In general I'd like to revert formatting changes that aren't relevant to actual code changes (is some editor doing that?). And maybe we can even clean away `checksumAddress` even further?

👍

**Comment by zacharyestep** [comment]
Any further changes?

File: data/basics/address.go [hunk 8]
```diff
@@ -70,44 +54,45 @@ func UnmarshalChecksumAddress(address string) (Address, error) {
 	}
 	var short Address
 	if len(decoded) < len(short) {
-		return Address{}, fmt.Errorf("decoded bad address: %v", address)
+		return Address{}, fmt.Errorf("decoded bad addr: %s %v", decoded, decoded)
```
Comment (Karmastic) [inline data/basics/address.go]: Should this be `, address, decoded`?

**Comment by Karmastic** [comment]
Thanks for pushing on this Zachary.

I see one change that might be an error, and have a question about impact
assessment.

File: cmd/goal/inspect.go [hunk 9]
```diff
@@ -72,7 +72,7 @@ type inspectTransaction struct {
 type inspectTxnHeader struct {
```
Comment (zeldovich) [inline cmd/goal/inspect.go]: With this change it looks like you should be able to remove `inspectTxnHeader` and `inspectPaymentTxnFields` altogether

File: data/basics/address.go [hunk 10]
```diff
@@ -70,44 +54,45 @@ func UnmarshalChecksumAddress(address string) (Address, error) {
...
+	incomingchecksum := decoded[len(decoded)-checksumLength:]
+
+	calculatedchecksum := short.GetChecksum()
+	isValid := bytes.Equal(incomingchecksum, calculatedchecksum)
 
-	checksumAddr := ChecksumAddress{short, decoded[len(decoded)-checksumLength:]}
-	if !checksumAddr.IsValid() {
-		return Address{}, fmt.Errorf("address %s is malformed, checksum verification failed", address)
+	if !isValid {
+		return Address{}, fmt.Errorf("address %s is malformed, checksum verification failed ", address)
```
Comment (zeldovich) [inline data/basics/address.go]: Why the added space at the end of the error message?

**Comment by zeldovich** [comment]
Is there any reason to keep `ChecksumAddress` in `daemon/algod/api/client/encoding.go`?

**Comment by zeldovich** [comment]
and `DigestChecksummed` from `cmd/auctionminion/encoding.go` for that matter

File: data/basics/address.go (hunk 8 above)
Comment (zacharyestep) [reply data/basics/address.go]: I think just address. I'll clean this up tonight.

**Comment by zacharyestep** [comment]
Karamstic, To quote the issue #51 - > This is unnecessary since we never represent an address as text without including the checksum, so `basics.Address` should just compute the checksum in its `String` and `MarshalText` methods (and verify the checksum in its `UnmarshalText` method). This will also render unnecessary the duplicate code in `cmd/goal/inspect.go`

I'll double-check that no parts of the codebase attempt to Unmarshal sans checksum, which is, I believe the state of affairs currently. and I will remove the extraneous checksumAddress types still lingering in the files identified by zeldovich. Thank you all.

File: daemon/algod/api/client/encoding.go [hunk 11]
```diff
@@ -60,5 +60,5 @@ func (a *ChecksumAddress) UnmarshalText(text []byte) error {
 // MarshalText implements the encoding.TextMarshaler interface
 func (a ChecksumAddress) MarshalText() (text []byte, err error) {
```
Comment (algoradam) [inline daemon/algod/api/client/encoding.go]: Any reason we need to keep this package's `ChecksumAddress` type or can we get rid of it too?

Comment (zacharyestep) [reply daemon/algod/api/client/encoding.go]: AFAIK, no. Will remove ASAP

**Comment by algoradam** [comment]
> The only concern I have that I need to consider is the possible impact of existing uses of Address.String() that are now going to be checksummed strings. Was this reviewed at all to understand the impact? If we have any cases where Address is persisted in string format, it will no longer decode/unmarshal correctly.
I believe it's safe. As far as I know, if we're using the non-checksummed string version of an address anywhere, then that should be considered a bug. Here are specific areas I've checked:

* Grepping the codebase for `UnmarshalText` doesn't show anywhere we explicitly try to unmarshal a non-checksummed address
* According to the go-codec documentation, (Un)marshalText is only used for text formats, so this shouldn't affect the behavior of `protocol.Encode` for msgpack encoding/decoding.
* When kmd uses addresses internally, it never converts them to strings; whenever kmd takes in an address via the API, it expects it to come with a checksum
* According to @rotemh when the REST APIs deal with addresses they always use the checksummed versions
* Grepping for `.String()`, I can only find one place we explicitly call `String()` on a non-checksummed address: a `log.Warn` statement in `data/accountManager.go:131` that should be using the checksummed version instead.

**Comment by algoradam** [comment]
Aside from the things @zeldovich and @karmastic already commented on, this looks good to me! Thank you for contributing!

**Comment by algobolson** [comment]
To test for explicit and implicit usages of Address.String() I checked out master, renamed Address.String() Address.StringTMP() and then looked at all the places that failed to compile. They were all log messages and using the checksum form is just as good or better because it's the format tools would want to take in if we copy-pasted that out of a log into something else. I then changed all the usages to StringTMP() and created String(){panic()} to and ran tests find implicit usages. The only thing was Address.MarshalText() ! So as long as the Marshal/Unmarshal test in address_test.go works I think it's safe to change the default behavior of Address.String()

**Comment by zacharyestep** [comment]
I'm using go 1.2.5 and it has slightly different opinions about what go fmt should/shouldn't do. Sorry if any more such changes sneak in I'll try hard not to let them.

File: cmd/auctionminion/main.go [hunk 12]
```diff
@@ -61,7 +61,7 @@ var stateFile = flag.String("statefile", "auctionminion.state", "State/config fi
 var initFlag = flag.Bool("init", false, "Initialize state file")
 
 type minionConfig struct {
-	AuctionKey DigestChecksummed
+	AuctionKey crypto.Digest
```
Comment (algoradam) [inline cmd/auctionminion/main.go]: Should be `basics.Address` rather than `crypto.Digest` if you want it to include the checksum.

Comment (zacharyestep) [reply cmd/auctionminion/main.go]: Agreed. Will fix tonight.

Comment (zacharyestep) [reply cmd/auctionminion/main.go]: Looks like this change broke a bunch of stuff in the auction area - I'm going to try to enlist the help of the relevant subject matter expert (Evan) for some "unsupervised fun" over the break to get it worked out. If that ends up taking a while would you consider taking this PR sans changes in the auction space?

Comment (Karmastic) [reply cmd/auctionminion/main.go]: Yes - feel free to submit an update without the auction changes and we'll re-evaluate it.

**Comment by zacharyestep** [comment]
Some thoughts. There seem to be a number of aliases to a crypto.Digest, Address.  If they all lived in the same package, wouldn't the compiler be able to reason that they are the same type at compile-time and they could be used interchangeably?

**Comment by derbear** [comment]
> Some thoughts. There seem to be a number of aliases to a crypto.Digest, Address. If they all lived in the same package, wouldn't the compiler be able to reason that they are the same type at compile-time and they could be used interchangeably? Can I just take the stuff in data/basics/address.go and move into crypto namespace to let the compiler be able to reason that they are interchangeable at runtime? Gotta review typing in go and the specifics of how aliases work across packages, but I think it's something like that. Are the different aliases given distinct responsibilities (seems like it).

The motivation behind making `crypto.Digest` and `Address` different types is to use Go's type-checker to catch certain kinds of bugs. For example, it would be bad if our code interpreted a 32-byte string as an address in one place and a block hash in another place--this would almost certainly mean that there's a severe safety bug somewhere.

I imagine that the Go compiler understands that these two defined types have the same underlying type (https://golang.org/ref/spec#Types) and will optimize appropriately (if that was your concern).

**Comment by zeldovich** [comment]
Can you delete the now-useless code from `cmd/goal/inspect.go`?  In particular, you should be able to delete `inspectTransaction`, `inspectTxnHeader`, `inspectPaymentTxnFields`, `txnToInspect`, and `txnFromInspect`.

File: cmd/goal/inspect.go (hunk 9 above)
Comment (algobolson) [reply cmd/goal/inspect.go]: I'm okay with that being a follow-on cleanup issue to get this big chunk of code in.

Comment (zeldovich) [reply cmd/goal/inspect.go]: Fair enough; we should just try to not lose track of it

Comment (algobolson) [reply cmd/goal/inspect.go]: noted in #129

**Pull request merged**